Overhaul cudf-polars docs for new streaming multi-GPU engines (#22252)
rapidsai/cudf: docs/cudf/source/cudf_polars/usage.md

(cudf-polars-usage)=
# Usage

`cudf-polars` enables GPU acceleration for Polars' `LazyFrame` API by executing logical plans
with cuDF. It requires only minimal code changes: create a GPU engine, then pass that engine to
each `collect()` call.

```python
import polars as pl
from cudf_polars.experimental.rapidsmpf.frontend.ray import RayEngine

with RayEngine() as engine:
    result = (
        pl.scan_parquet("/data/dataset/*.parquet")
        .filter(pl.col("amount") > 100)
        .group_by("customer_id")

>         .agg(pl.col("amount").sum())

Reviewer: Above, we say "minimal code changes". To reinforce this, move the query construction outside the context manager:

```
query = (
        pl.scan_parquet("/data/dataset/*.parquet")
        .filter(pl.col("amount") > 100)
        .group_by("customer_id")
        .agg(pl.col("amount").sum())
)

with RayEngine() as engine:
    result = query.collect(engine=engine)
```

        .collect(engine=engine)
    )
    print(result)
```
In the example above, we use {class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine`,
which is the preferred engine for GPU acceleration. Other engines are also available, including
{doc}`DaskEngine <dask_engine>` and {doc}`SPMDEngine <spmd_engine>`. All of these engines provide
multi-GPU and multi-node execution. They partition inputs and stream data through the query
graph, allowing execution to scale beyond device memory and across multiple GPUs out of the box.

> For the most lightweight streaming setup, a single GPU with no Ray, Dask, or `rrun` launcher,

Reviewer: ```suggestion
For the simplest streaming setup, a single GPU with no Ray, Dask, or `rrun` launcher,
```

> see the [Single-GPU setup](spmd_engine.md#single-gpu-setup).

Reviewer: ```suggestion
see [Single-GPU setup](spmd_engine.md#single-gpu-setup).
```

> As a separate alternative, `cudf-polars` also provides an *in-memory* engine. This engine uses a

Reviewer: ```suggestion
As an alternative, `cudf-polars` also provides an *in-memory* engine. This engine uses a
```

simpler execution model: each Polars operation is translated into a corresponding GPU operation

> and executed on a single GPU. See [Polars GPU Support](https://docs.pola.rs/user-guide/gpu-support/).

Reviewer: Readers will naturally ask when I should use streaming vs. in-memory here. We should mention that the in-memory engine works fine for smaller datasets, but won't scale to datasets that are larger than GPU memory.

## Configuring `RayEngine`

> {class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine` with no arguments uses

Reviewer: `RayEngine` isn't the the API reference yet.

Author: I think we should keep the ref so that it will work when it gets in?

> built-in defaults: it calls `ray.init()` to start a local [Ray][ray-docs] cluster and creates one

Reviewer: "with no arguments uses built-in defaults" is redundant :)

Maybe just

> `RayEngine` starts a local Ray cluster and creates one GPU worker per visible GPU.

GPU worker per visible GPU.

For custom configuration, build a
{class}`~cudf_polars.experimental.rapidsmpf.frontend.options.StreamingOptions` and use
`RayEngine.from_options()`:

```python
import polars as pl
from cudf_polars.experimental.rapidsmpf.frontend.options import StreamingOptions
from cudf_polars.experimental.rapidsmpf.frontend.ray import RayEngine

opts = StreamingOptions(num_streaming_threads=8, fallback_mode="silent")

with RayEngine.from_options(opts) as engine:
    result = pl.scan_parquet("/data/dataset/*.parquet").collect(engine=engine)
```

See {doc}`options` for the available fields.

```{note}

> {class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine` is **instance-based**: you

Reviewer: I'm not sure I know what "instance-based" means here.

must create one and hand it to `.collect(engine=engine)`. Prefer the context-manager form so the
Ray cluster and GPU workers are torn down automatically.
```

## Attaching to an existing Ray cluster

> For multi-machine runs, start a Ray cluster separately (for example with `ray start` on each

Reviewer: We say "multi-node" earlier, but "multi-machine" here. Let's pick one (no preference which).

node) and attach to it from your driver script. When Ray is already initialized,
{class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine` connects to the running
cluster and leaves it untouched on exit:

```python
import ray
import polars as pl
from cudf_polars.experimental.rapidsmpf.frontend.ray import RayEngine

ray.init(address="auto")  # attach to a running cluster
with RayEngine() as engine:
    result = (
        pl.scan_parquet("s3://bucket/*.parquet")
            .group_by("customer_id")
            .agg(pl.col("amount").sum())
            .collect(engine=engine)
    )
```

{class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine` creates one rank per GPU in
the Ray cluster and bootstraps a UCXX communicator across them. It raises `RuntimeError` if
created inside an `rrun` cluster or if no GPUs are available.

> ## Using `RayEngine` in a Jupyter notebook

Reviewer: FWIW, I'd rather give a name to this style of usage, and then mention that it's useful in contexts like a Jupyter notebook.

Maybe something call the section "Manually Engine Lifetime Control"?

In a notebook, the context-manager form is inconvenient because a `with` block cannot span
multiple cells — the engine would be torn down at the end of the cell that created it. Instead,
construct {class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine` once and reuse it
across cells, then call `engine.shutdown()` when you are done:

```python
# Cell 1: start the engine
from cudf_polars.experimental.rapidsmpf.frontend.ray import RayEngine

engine = RayEngine()
```

```python
# Cell 2: run a query
import polars as pl

result = (
    pl.scan_parquet("/data/*.parquet")
      .group_by("customer_id")
      .agg(pl.col("amount").sum())
      .collect(engine=engine)
)
result
```

```python
# Cell 3: run another query reusing the same engine
other = pl.scan_parquet("/data/other/*.parquet").collect(engine=engine)
```

```python
# Final cell: tear everything down
engine.shutdown()
```

> `engine.shutdown()` stops the rank actors and, if the engine started Ray itself, also calls

Reviewer: "rank actors" hasn't really been introduced yet. It's used here and under "Cluster diagnostics".

`ray.shutdown()`. It is idempotent, so calling it twice is safe.

## Cluster diagnostics

> {meth}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine.gather_cluster_info` returns

Reviewer: The output (or at least its schema) could be useful here. If it's in the docstring of `gather_cluster_info` then that's fine.

(cudf-polars-usage)=
# Usage

`cudf-polars` runs your Polars `LazyFrame` queries on GPU. You select GPU execution by passing
an `engine=` argument to `.collect()` or `.sink_*()`. See {doc}`engines` for the conceptual
picture; this page walks through running your first query.

> We recommend constructing an engine explicitly even on a single GPU. The engine constructor is

Reviewer 2: ```suggestion
We always recommend constructing an engine object and using them in a context manager to ensure proper resource cleanup. The engine constructor is
```

such as `spill_to_pinned_memory` or `fallback_mode`. Ray is the showcased example below; see also
{doc}`other_engines`.

## Your first GPU query

```python
import polars as pl
from cudf_polars.experimental.rapidsmpf.frontend.ray import RayEngine

query = (
    pl.scan_parquet("/data/dataset/*.parquet")
    .filter(pl.col("amount") > 100)
    .group_by("customer_id")
    .agg(pl.col("amount").sum())
)

with RayEngine() as engine:
    result = query.collect(engine=engine)
print(result)
```

{class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine` with no arguments uses every
GPU visible to the process, so the example above runs on one GPU if that's all that's available
and scales automatically to every GPU on the node otherwise. It also attaches to an existing
Ray cluster if one is already running (see [Attaching to an existing Raycluster](#attaching-to-an-existing-ray-cluster)).

```{note}
The examples on this page use {class}`~cudf_polars.experimental.rapidsmpf.frontend.ray.RayEngine`. `cudf-polars` supports
multiple engines for GPU execution. See {doc}`other_engines` for alternatives, or {doc}`engines` for a conceptual overview of when to pick which.
```

> ## Configuring `RayEngine`

Reviewer 2: suggestion: The next few points appear very RayEngine specific. Since you also created a `dask_engine.md` page, would it be worth delegating all this detail to `ray_engine.md`?

Author: For this release, the focus is the Ray engine, so I think it is reasonable for the main usage page to lean more heavily into `RayEngine`-specific details for now.

    .filter(pl.col("amount") > 100)
    .group_by("customer_id")
    .agg(pl.col("amount").sum())
)

with RayEngine() as engine:
    result = query.collect(engine=engine)
print(result)
```

{class}`~cudf_polars.engine.ray.RayEngine` with no arguments uses every
GPU visible to the process, so the example above runs on one GPU if that's all that's available
and scales automatically to every GPU on the node otherwise. It also attaches to an existing
Ray cluster if one is already running (see [Attaching to an existing Raycluster](#attaching-to-an-existing-ray-cluster)).

```{note}
The examples on this page use {class}`~cudf_polars.engine.ray.RayEngine`. `cudf-polars` supports
multiple engines for GPU execution. See {doc}`other_engines` for alternatives, or {doc}`engines` for a conceptual overview of when to pick which.
```

```{note}
`.collect()` pulls the full result back to the driver process. For large distributed outputs,
prefer `.sink_*()` or aggregate/sample inside the query before `.collect()`. See
[Result collection](engines.md#result-collection).
```

## Configuring `RayEngine`

The same `from_options()` / `StreamingOptions` pattern shown here works for every streaming

> engine — see {doc}`other_engines` for the DaskEngine and SPMDEngine variants.

Reviewer 3: ```suggestion
engine; see {doc}`other_engines` for the DaskEngine and SPMDEngine variants.
```

Reviewer 3: Also, I think move this paragraph to the end of the section, it's getting in the way of the user running something with `RayEngine`...

    result = pl.scan_parquet("/data/dataset/*.parquet").collect(engine=engine)
```

See {doc}`options` for the available fields.

```{note}
`RayEngine` is an object you create and pass to `.collect(engine=engine)`. Prefer the
context-manager form so the Ray cluster and GPU workers are torn down automatically.
```

## Attaching to an existing Ray cluster

For multi-node runs, start a Ray cluster separately (for example with `ray start` on each
node) and attach to it from your driver script. When Ray is already initialized,
{class}`~cudf_polars.engine.ray.RayEngine` connects to the running
cluster and leaves it untouched on exit:

```python
import ray
import polars as pl
from cudf_polars.engine.ray import RayEngine

ray.init(address="auto")  # attach to a running cluster
with RayEngine() as engine:
    result = (
        pl.scan_parquet("s3://bucket/*.parquet")
            .group_by("customer_id")
            .agg(pl.col("amount").sum())
            .collect(engine=engine)
    )
```

{class}`~cudf_polars.engine.ray.RayEngine` creates one rank per GPU in
the Ray cluster and bootstraps a UCXX communicator across them. It raises `RuntimeError` if

> created inside an `rrun` cluster or if no GPUs are available.

Reviewer 3: This isn't part of a howto, I think. The user has never heard of `rrun` up to now so why would they care.

Author: Good point, removed, now we just have
```
{class}`~cudf_polars.engine.ray.RayEngine` creates one rank per GPU in 
the Ray cluster. It raises `RuntimeError` if no GPUs are available.
```
